File: glean/lang/kotlin/indexer/glean_utils/Type.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

package glean.lang.kotlin.indexer.glean_utils

import com.facebook.glean.schema.javakotlin_alpha.Type as JavaKotlinType
import com.facebook.glean.schema.kotlin_alpha.Type
import com.facebook.glean.schema.kotlin_alpha.TypeArg
import com.facebook.glean.schema.kotlin_alpha.TypeArgKey
import com.facebook.glean.schema.kotlin_alpha.TypeKey
import org.jetbrains.kotlin.backend.jvm.ir.psiElement
import org.jetbrains.kotlin.resolve.BindingContext
import org.jetbrains.kotlin.types.KotlinType
import org.jetbrains.kotlin.types.typeUtil.TypeNullability
import org.jetbrains.kotlin.types.typeUtil.nullability

private fun KotlinType.javakotlinType(): JavaKotlinType {
  return JavaKotlinType.Builder().setKey(this.gleanType()).build()
}

private fun KotlinType.asTypeArg(bindingContext: BindingContext): TypeArg {
  val key = TypeArgKey.Builder().setType(buildKotlinType(this, bindingContext)).build()
  return TypeArg.Builder().setKey(key).build()
}

fun buildKotlinType(typeReference: KotlinType, bindingContext: BindingContext): Type {
  val typeKey = TypeKey.Builder()
  typeKey.type = typeReference.javakotlinType()
  val typeConstructorElement = typeReference.constructor.declarationDescriptor?.psiElement
  if (typeConstructorElement != null) {
    typeKey.location = buildFileLocation(typeConstructorElement)
  }

  val typeArgs: MutableList<TypeArg> = mutableListOf()
  for (typeArg in typeReference.arguments) {
    if (typeArg.isStarProjection) {
      continue
    }
    typeArgs.add(typeArg.type.asTypeArg(bindingContext))
  }
  typeKey.typeArgs = typeArgs

  typeKey.isIsNullable = typeReference.nullability() == TypeNullability.NULLABLE
  return Type.Builder().setKey(typeKey.build()).build()
}

File: glean/lang/csharp/indexer/Schema/CSharp/Type.cs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

using Microsoft.CodeAnalysis;
using Serilog;

namespace Glean.Indexer.Schema.CSharp;

public record Type(NamedType? NamedType = default, TypeParameterFact? TypeParameter = default, ArrayTypeFact? ArrayType = default)
{
    public static bool TryFromSymbol(ITypeSymbol symbol, out Type? result)
    {
        switch (symbol) {
            case INamedTypeSymbol namedTypeSymbol:
                if (NamedType.TryFromSymbol(namedTypeSymbol, out var namedType) && namedType != null)
                {
                    result = new Type(NamedType: namedType);
                    return true;
                }
                break;
            case ITypeParameterSymbol typeParameterSymbol:
                var typeParameter = TypeParameterFact.FromSymbol(typeParameterSymbol);

                result = new Type(TypeParameter: typeParameter);
                return true;
            case IArrayTypeSymbol arrayTypeSymbol:
                if (ArrayTypeFact.TryFromSymbol(arrayTypeSymbol, out var arrayType) && arrayType != null)
                {
                    result = new Type(ArrayType: arrayType);
                    return true;
                }
                break;
            default:
                Log.Error($"Unsupported type {symbol}");
                break;
        }

        result = default;
        return false;
    }
}
